pkg/pubsub/system.go

```diff
@@ -0,0 +1,89 @@
+package pubsub
+
+import (
+	"context"
+	"sync"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/connection"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/provider"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("pubsub-system")
+
+type System struct {
+	mux         sync.RWMutex
+	connections map[string]connection.Connection
+}
+
+func NewSystem() *System {
+	return &System{}
+}
+
+func (p *System) Publish(ctx context.Context, connection string, topic string, msg interface{}) error {
+	p.mux.RLock()
+	defer p.mux.RUnlock()
+	if len(p.connections) > 0 {
+		if c, ok := p.connections[connection]; ok {
+			return c.Publish(context.Background(), msg, topic)
+		}
+		log.Info("Connection is not initialized", "name", connection)
+	} else {
+		log.Info("No pub sub tools are enabled")
+	}
+	return nil
+}
+
+func (p *System) PublishBatch(ctx context.Context, connection string, topic string, msg []interface{}) error {
+	p.mux.RLock()
+	defer p.mux.RUnlock()
+	if len(p.connections) > 0 {
+		if c, ok := p.connections[connection]; ok {
+			return c.PublishBatch(context.Background(), msg, topic)
+		}
+		log.Info("Connection is not initialized", "name", connection)
+	} else {
+		log.Info("No pub sub tools are enabled")
+	}
+	return nil
+}
+
+func (p *System) InitiateConnection(ctx context.Context, newConnection provider.InitiateConnection, config interface{}, name string) error {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+	tmp, err := newConnection(ctx, config)
+	if err != nil {
+		return err
+	}
+	if p.connections == nil {
+		p.connections = map[string]connection.Connection{}
+	}
+	p.connections[name] = tmp
+
+	return nil
+}
+
+func (p *System) CloseConnection(connection string) error {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+	if len(p.connections) > 0 {
+		if c, ok := p.connections[connection]; ok {
+			err := c.CloseConnection()
+			if err != nil {
+				return err
+			}
+			delete(p.connections, connection)
+		}
+	}
+	return nil
+}
+
+func (p *System) IsBatchingEnabled(connection string) bool {
```

Comment:
If callers need to be aware of the capabilities of a specific connection, then `system` is failing in its job as an abstraction. This function should not exist.

Reply:
resolving this conversation in favor of raising a follow up PR for batch publish changes.